File: utcp-client-mcp.ts
// utcp-client-mcp.ts (TypeScript MCP-UTCP Bridge Server)

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import path from "path";
import { promises as fs } from "fs";
import { parse as parseDotEnv } from 'dotenv';

import {
    UtcpClient,
    CallTemplateSchema,
    InMemConcurrentToolRepository,
    TagSearchStrategy,
    DefaultVariableSubstitutor,
    ensureCorePluginsInitialized
} from "@utcp/sdk";
import type { UtcpClientConfig } from "@utcp/sdk";

ensureCorePluginsInitialized();

let utcpClient: UtcpClient | null = null;

async function main() {
    console.log("Initializing UTCP-MCP Bridge...");
    setupMcpTools();
    utcpClient = await initializeUtcpClient();

    const bridgeScriptPath = path.resolve(import.meta.dir, import.meta.file);
    const connectionConfig = {
        mcpServers: {
            "typescript-utcp-bridge": {
                command: "bun",
                args: ["run", bridgeScriptPath],
            }
        }
    };

    console.log("\n✅ Bridge is ready. To connect, use this configuration in your MCP client's config file:");
    console.log("================================ MCP CONFIG ================================");
    console.log(JSON.stringify(connectionConfig, null, 2));
    console.log("==========================================================================");

    console.log("\nStarting MCP server on stdio...");
    const transport = new StdioServerTransport();
    await mcp.connect(transport);
}

const mcp = new McpServer({
    name: "UTCP-Client-MCP-Bridge",
    version: "1.0.0",
});

function setupMcpTools() {
    mcp.registerTool("register_manual", {
        title: "Register a UTCP Manual",
        description: "Registers a new tool provider by providing its call template.",
        inputSchema: { manual_call_template: CallTemplateSchema.describe("The call template for the UTCP Manual endpoint.") },
    }, async (input) => {
        const client = await initializeUtcpClient();
        try {
            const result = await client.registerManual(input.manual_call_template as any);
            return { content: [{ type: "text", text: JSON.stringify(result) }] };
        } catch (e: any) {
            return { content: [{ type: "text", text: JSON.stringify({ success: false, error: e.message }) }] };
        }
    });

    mcp.registerTool("deregister_manual", {
        title: "Deregister a UTCP Manual",
        description: "Deregisters a tool provider from the UTCP client.",
        inputSchema: { manual_name: z.string().describe("The name of the manual to deregister.") },
    }, async (input) => {
        const client = await initializeUtcpClient();
        try {
            const success = await client.deregisterManual(input.manual_name);
            const message = success ? `Manual '${input.manual_name}' deregistered.` : `Manual '${input.manual_name}' not found.`;
            return { content: [{ type: "text", text: JSON.stringify({ success, message }) }] };
        } catch (e: any) {
            return { content: [{ type: "text", text: JSON.stringify({ success: false, error: e.message }) }] };
        }
    });

    mcp.registerTool("call_tool", {
        title: "Call a UTCP Tool",
        description: "Calls a registered tool by its full namespaced name.",
        inputSchema: {
            tool_name: z.string().describe("The full name of the tool to call."),
            arguments: z.record(z.string(), z.any()).describe("A JSON object of arguments."),
        },
    }, async (input) => {
        const client = await initializeUtcpClient();
        try {
            const result = await client.callTool(input.tool_name, input.arguments);
            return { content: [{ type: "text", text: JSON.stringify({ success: true, tool_name: input.tool_name, result }) }] };
        } catch (e: any) {
            return { content: [{ type: "text", text: JSON.stringify({ success: false, error: e.message }) }] };
        }
    });

    mcp.registerTool("search_tools", {
        title: "Search for UTCP Tools",
        description: "Searches for relevant tools based on a task description.",
        inputSchema: {
            task_description: z.string().describe("A natural language description of the task."),
            limit: z.number().optional().default(10),
        },
    }, async (input) => {
        const client = await initializeUtcpClient();
        try {
            const tools = await client.searchTools(input.task_description, input.limit);
            const simplified = tools.map(t => ({ name: t.name, description: t.description, input_schema: t.inputs }));
            return { content: [{ type: "text", text: JSON.stringify({ tools: simplified }) }] };
        } catch (e: any) {
            return { content: [{ type: "text", text: JSON.stringify({ success: false, error: e.message }) }] };
        }
    });

    mcp.registerTool("list_tools", {
        title: "List All Registered UTCP Tools",
        description: "Returns a list of all tools currently registered.",
        inputSchema: {},
    }, async () => {
        const client = await initializeUtcpClient();
        try {
            const tools = await client.config.tool_repository.getTools();
            const toolInfo = tools.map(t => ({ name: t.name, description: t.description }));
            return { content: [{ type: "text", text: JSON.stringify({ tools: toolInfo }) }] };
        } catch (e: any) {
            return { content: [{ type: "text", text: JSON.stringify({ success: false, error: e.message }) }] };
        }
    });
}

async function initializeUtcpClient(): Promise<UtcpClient> {
    if (utcpClient) {
        return utcpClient as UtcpClient;
    }

    const scriptDir = path.resolve(import.meta.dir);
    const configPath = path.resolve(scriptDir, '.utcp_config.json');
    console.log(`Searching for UTCP config file at: ${configPath}`);

    let rawConfig: any = {};
    try {
        const configFileContent = await fs.readFile(configPath, 'utf-8');
        rawConfig = JSON.parse(configFileContent);
        console.log("Loaded UTCP client configuration from .utcp_config.json");
    } catch (e: any) {
        if (e.code !== 'ENOENT') {
            console.warn(`Could not read or parse .utcp_config.json. Error: ${e.message}`);
        }
        console.log("No valid .utcp_config.json found. Initializing with default config.");
    }

    const toolRepository = new InMemConcurrentToolRepository();
    const searchStrategy = new TagSearchStrategy({
        tool_search_strategy_type: 'tag_and_description_word_match',
        ...rawConfig.tool_search_strategy
    });
    const variableSubstitutor = new DefaultVariableSubstitutor();

    const loadedVariables: Record<string, string> = {};
    if (rawConfig.load_variables_from) {
        for (const loaderConfig of rawConfig.load_variables_from) {
            if (loaderConfig.variable_loader_type === 'dotenv' && loaderConfig.env_file_path) {
                try {
                    const envPath = path.resolve(scriptDir, loaderConfig.env_file_path);
                    const envContent = await fs.readFile(envPath, 'utf-8');
                    Object.assign(loadedVariables, parseDotEnv(envContent));
                    console.log(`Successfully loaded variables from ${envPath}`);
                } catch (e: any) {
                    console.warn(`Could not load .env file from '${loaderConfig.env_file_path}': ${e.message}`);
                }
            }
        }
    }

    const finalVariables = { ...loadedVariables, ...(rawConfig.variables || {}) };

    const clientConfig: UtcpClientConfig = {
        variables: finalVariables,
        load_variables_from: [],
        tool_repository: toolRepository,
        tool_search_strategy: searchStrategy,
        post_processing: rawConfig.post_processing || [],
        manual_call_templates: [],
    };

    const newClient = new (UtcpClient as any)(clientConfig, variableSubstitutor, scriptDir);

    if (rawConfig.manual_call_templates && Array.isArray(rawConfig.manual_call_templates)) {
        console.log(`Registering ${rawConfig.manual_call_templates.length} initial manuals from config...`);
        for (const manualTemplate of rawConfig.manual_call_templates) {
            try {
                if (!manualTemplate.name) {
                    manualTemplate.name = manualTemplate.call_template_type || 'unnamed_manual';
                }
                await newClient.registerManual(manualTemplate);
            } catch (e: any) {
                console.error(`Failed to register initial manual '${manualTemplate.name}': ${e.message}`);
            }
        }
    }

    console.log("UTCP Client manually initialized successfully.");
    utcpClient = newClient;
    return utcpClient as UtcpClient;
}

main().catch(err => {
    console.error("Failed to start UTCP-MCP Bridge:", err);
    process.exit(1);
});

File: python_mcp_bridge/utcp-client-mcp.py
# /// script
# dependencies = [
#   "fastmcp",
#   "utcp==1.0.1",
#   "utcp-mcp==1.0.1",
#   "utcp-text==1.0.1",
#   "utcp-cli==1.0.1",
#   "utcp-http==1.0.1",
# ]
# ///
"""FastMCP stdio server that proxies UTCP client functionalities as tools.

This server provides MCP tools that expose the core UTCP client operations:
- Registering and deregistering tool providers
- Calling tools through providers
- Searching for available tools
- Loading providers from a JSON configuration file
- Getting required variables for tools

The server automatically loads providers from a 'providers.json' file in the same directory.
"""

import asyncio
import json
from pathlib import Path
from typing import Any, Dict, Optional

from fastmcp import FastMCP
from pydantic import BaseModel, Field

# UTCP imports
from utcp.utcp_client import UtcpClient
from utcp.data.utcp_client_config import UtcpClientConfig
from utcp.data.call_template import CallTemplate

# Global UTCP client instance
utcp_client: Optional[UtcpClient] = None

# Initialize FastMCP server
mcp = FastMCP("UTCP Client MCP Server")

async def initialize_utcp_client():
    """Initialize the UTCP client and try to load providers.json from the same directory."""
    global utcp_client
    
    if utcp_client is not None:
        return utcp_client
    
    script_dir = Path(__file__).parent
    config_file = script_dir / ".utcp_config.json"
    
    if config_file.exists():
        config = json.loads(config_file.read_text())
        config = UtcpClientConfig.model_validate(config)
    else:
        config = UtcpClientConfig()

    # Create UTCP client
    utcp_client = await UtcpClient.create(config=config)
    
    return utcp_client


@mcp.tool()
async def register_manual(manual_call_template: CallTemplate) -> Dict[str, Any]:
    """Register a new tool provider with the UTCP client.
    
    Args:
        manual_call_template: Call template to the endpoint of a UTCP Manual
        
    Returns:
        Dictionary with success status and list of registered tools
    """
    client = await initialize_utcp_client()
    
    try:
        tools = await client.register_manual(manual_call_template)
        
        return {
            "success": True,
            "manual_name": manual_call_template.name,
            "tools_registered": len(tools),
            "tool_names": [tool.name for tool in tools]
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


@mcp.tool()
async def deregister_manual(manual_name: str) -> Dict[str, Any]:
    """Deregister a tool provider from the UTCP client.
    
    Args:
        manual_name: Name of the manual to deregister
        
    Returns:
        Dictionary with success status
    """
    client = await initialize_utcp_client()
    
    try:
        await client.deregister_manual(manual_name)
        return {
            "success": True,
            "message": f"Manual '{manual_name}' deregistered successfully"
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


@mcp.tool()
async def call_tool(tool_name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
    """Call a tool through the UTCP client.
    
    Args:
        tool_name: Name of the tool to call
        arguments: Arguments for the tool call
        
    Returns:
        Dictionary with success status and tool result
    """
    client = await initialize_utcp_client()
    
    try:
        result = await client.call_tool(tool_name, arguments)
        return {
            "success": True,
            "tool_name": tool_name,
            "result": result
        }
    except Exception as e:
        return {
            "success": False,
            "tool_name": tool_name,
            "error": str(e)
        }


@mcp.tool()
async def search_tools(task_description: str, limit: int = 10) -> Dict[str, Any]:
    """Search for tools using a query string.
    
    Args:
        task_description: Description of the task to search for tools
        limit: Optional limit on the number of tools to return
        
    Returns:
        Dictionary with success status and matching tools
    """
    client = await initialize_utcp_client()
    
    try:
        tools = await client.search_tools(task_description, limit)
        return {"tools": [{"name": tool.name, "description": tool.description, "input_schema": tool.inputs.model_dump(exclude_none=True)} for tool in tools]}
    except Exception as e:
        return {"error": str(e)}


@mcp.tool()
async def get_required_keys_for_tool(tool_name: str) -> Dict[str, Any]:
    """Get required environment variables for a registered tool.
    
    Args:
        tool_name: Name of the tool to get variables for
        
    Returns:
        Dictionary with success status and required variables
    """
    client = await initialize_utcp_client()
    
    try:
        variables = await client.get_required_variables_for_registered_tool(tool_name)
        return {
            "success": True,
            "tool_name": tool_name,
            "required_variables": variables
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }


@mcp.tool()
async def tool_info(tool_name: str) -> Dict[str, Any]:
    """Get complete information about a specific tool including all details using model_dump().
    
    Args:
        tool_name: Name of the tool to get complete information for
        
    Returns:
        Dictionary with success status and complete tool information with model_dump()
    """
    client = await initialize_utcp_client()
    
    try:
        # Search for the specific tool
        tool = await client.config.tool_repository.get_tool(tool_name)
        
        if not tool:
            return {
                "success": False,
                "error": f"Tool '{tool_name}' not found"
            }
        
        # Return complete tool information with model_dump()
        return {
            "success": True,
            "tool": tool.model_dump()
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e)
        }

async def main():
    global utcp_client
    utcp_client = await initialize_utcp_client()
    await mcp.run_async(transport="stdio")

if __name__ == "__main__":
    # Run the FastMCP server
    asyncio.run(main())
